File: app/src/main/java/com/tutv/android/ui/home/HomePresenter.java
package com.tutv.android.ui.home;

import com.tutv.android.domain.Genre;
import com.tutv.android.repository.SeriesRepository;
import com.tutv.android.utils.schedulers.BaseSchedulerProvider;

import java.lang.ref.WeakReference;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;

public class HomePresenter {

    private final WeakReference<HomeView> view;
    private final SeriesRepository seriesRepository;
    private final BaseSchedulerProvider schedulerProvider;

    private final CompositeDisposable disposables;

    public HomePresenter(HomeView view, SeriesRepository seriesRepository, BaseSchedulerProvider schedulerProvider) {
        this.view = new WeakReference<>(view);
        this.seriesRepository = seriesRepository;
        this.schedulerProvider = schedulerProvider;

        this.disposables = new CompositeDisposable();
    }

    public void onViewAttached() {
        disposables.add(seriesRepository.getGenres()
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::genresLoadSuccessful, this::genresLoadError)
        );
    }

    public void onViewDetached() {
        disposables.dispose();
    }

    private void genresLoadSuccessful(List<Genre> genres) {
        for(Genre genre : genres)
            if(view.get() != null)
                view.get().createGenreList(genre.getId(), genre.getName());
    }

    private void genresLoadError(Throwable e) {
        HomeView actualView = view.get();
        if (actualView != null)
            actualView.showError();
    }
}

File: app/src/main/java/com/tutv/android/ui/home/HomePresenter.kt
package com.tutv.android.ui.home

import com.tutv.android.domain.Genre
import com.tutv.android.repository.SeriesRepository
import com.tutv.android.utils.schedulers.BaseSchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

class HomePresenter(
        view: HomeView,
        var seriesRepository: SeriesRepository,
        var schedulerProvider: BaseSchedulerProvider) {

    private val view: WeakReference<HomeView?> = WeakReference(view)
    private val disposables: CompositeDisposable = CompositeDisposable()

    fun onViewAttached() {
        disposables.add(seriesRepository.genres
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe({ genres: List<Genre> -> genresLoadSuccessful(genres) }) { e: Throwable? -> genresLoadError(e) }
        )
    }

    fun onViewDetached() {
        disposables.dispose()
    }

    private fun genresLoadSuccessful(genres: List<Genre>) {
        for (genre in genres) view.get()?.createGenreList(genre.id, genre.name)
    }

    private fun genresLoadError(e: Throwable?) {
        val actualView = view.get()
        actualView?.showError()
    }

}